### ISSUE TITLE
ExtendVirtualDisk succeeds, but the capacity of the disk does not change.

### ISSUE THREAD

**Issue by unclethan** [vmware/pyvmomi#132]
I attempted to change the size of a virtual disk via: 
si.content.virtualDiskManager.ExtendVirtualDisk(....)

The task reported as successful (as task.info.state reported, as did the vSphere web client tasks list). However looking at the Virtual Machine that uses the disk, the capacity has not changed according to vSphere or pyVmomi. Is there a step I am missing to have the virtual machine pick up the new disk capacity?

See https://github.com/vmware/pyvmomi/issues/119

**Comment by mekalag** [comment]
Hi, 

I have used ReconfigVM_Task(spec=spec), when I click on VM and select Edit settings, I see new hard disk added. 

But in Disk management this Hard disk size shows unallocated. How do i create new volume in the machine , say E drive. 

which function to call in PyVmomi. Kindly suggest

**Comment by purple4reina** [comment]
PyVmomi can extend the size of the disk, but the OS will not yet recognize it. You will need to make a separate call to do that.

**Comment by mekalag** [comment]
Thank you. 
I wanted to know if there are any PyVmomi functions to do extend size of the disk on the machine.
Was looking for any functions which could reflect the new drive on VM.

**Comment by michaelrice** [comment]
The process is not the same on all supported platforms. You could use the GuestOperations to log into the VM and issue the command needed for the given platform..

**Comment by zhongshun89** [comment]
the disksapce you expand can't list by “df -hl”。use “fdisk -l” ,you can se the expand space dev。

**Comment by zhongshun89** [comment]
my code：
    def expendDiskCapacity(self, diskKey, newCapacityInGB):
        errorMsg = None
        deviceChange = []
        for diskDeviceItem in self.vmHost.config.hardware.device:
            if isinstance(diskDeviceItem, vim.vm.device.VirtualDisk) and diskDeviceItem.key == diskKey:
                oldDiskCapacity = diskDeviceItem.capacityInKB
                newDiskCapacity = newCapacityInGB*1024*1024
                if(newDiskCapacity > oldDiskCapacity):
                    capacitySpec = vim.vm.device.VirtualDeviceSpec()
                    capacitySpec.operation = vim.vm.device.VirtualDeviceSpec.Operation.edit
                    capacitySpec.device = diskDeviceItem
                    capacitySpec.device.capacityInKB = newDiskCapacity
                    # 因为不能处理中文字符比编码的问题，这里置为空，其实没有修改这个值
                    capacitySpec.device.deviceInfo.label = ""
                    deviceChange.append(capacitySpec)
        if len(deviceChange) > 0:
            vmconfig = vim.vm.ConfigSpec()
            vmconfig.deviceChange = deviceChange
            resetTask = self.vmHost.ReconfigVM_Task(vmconfig)
            tasks.wait_for_tasks(self.vmConnect, [resetTask])
            if resetTask.info.state == 'error':
                errorMsg = "扩展磁盘空间出错:" + resetTask.info.error.msg
        return errorMsg

**Comment by zhongshun89** [comment]
def expendDiskCapacity(self, diskKey, newCapacityInGB):
        errorMsg = None
        deviceChange = []
        for diskDeviceItem in self.vmHost.config.hardware.device:
            if isinstance(diskDeviceItem, vim.vm.device.VirtualDisk) and diskDeviceItem.key == diskKey:
                oldDiskCapacity = diskDeviceItem.capacityInKB
                newDiskCapacity = newCapacityInGB*1024*1024
                if(newDiskCapacity > oldDiskCapacity):
                    capacitySpec = vim.vm.device.VirtualDeviceSpec()
                    capacitySpec.operation = vim.vm.device.VirtualDeviceSpec.Operation.edit
                    capacitySpec.device = diskDeviceItem
                    capacitySpec.device.capacityInKB = newDiskCapacity
                    # 因为不能处理中文字符比编码的问题，这里置为空，其实没有修改这个值
                    capacitySpec.device.deviceInfo.label = ""
                    deviceChange.append(capacitySpec)
        if len(deviceChange) > 0:
            vmconfig = vim.vm.ConfigSpec()
            vmconfig.deviceChange = deviceChange
            resetTask = self.vmHost.ReconfigVM_Task(vmconfig)
            tasks.wait_for_tasks(self.vmConnect, [resetTask])
            if resetTask.info.state == 'error':
                errorMsg = "扩展磁盘空间出错:" + resetTask.info.error.msg
        return errorMsg

**Comment by diecav** [comment]
Hi all
I’m trying to manage this topic to.
Just for summarize, increase the Disk size is not a big thing. Just put the new Disk size as mentioned by @whereismyjetpack 

`disk_spec.device.capacityInKB = new_disk_kb`

then use ReconfigVM_Task.

But what about expand the partition inside the increased disk?
Thes are some parameters,methods,options...?

In Powershell they provide a cmdled thet can be used like that:
`Set-HardDisk -HardDisk $disk -CapacityGB 100 -GuestUser Administrator -GuestPassword Password -Partition "C" -ResizeGuestPartition -Confirm:$false`

Someone find something similar or the only one way is to use the vim.vm.guest.ProcessManager and perform specific OS calls?

**Comment by hartsock** [comment]
@diecav as you mention, this is highly OS specific ... each OS will require it's own calls/programs to recognize the new space and grow its file systems to match. Some people may prefer to have the new free space mapped as a new file system entirely. So it's all highly subjective.

IE:
* windows http://www.howtogeek.com/howto/windows-vista/resize-a-partition-for-free-in-windows-vista/
* Ubuntu http://askubuntu.com/questions/126153/how-to-resize-partitions
* CentOS https://www.centos.org/docs/5/html/Cluster_Logical_Volume_Manager/lv_extend.html

... I could keep going ...

The phrase to use with Google is [grow file system to fill free space on disk](https://www.google.com/search?q=grow+file+system+to+fill+free+space+on+disk) and then add your particular OS. You'll then have to come up with a way to call this via your Guest tools.

**tl;dr** this is a very broad topic and we could have a whole thread on each guest OS. Instead, use the samples, google, and Guest Tools to fabricate your specific answer yourself. You're of course encouraged to submit a sample for this if you would like but it's a very narrow focus.

Good luck! :-)

**Comment by avinash321** [comment]
Hi, can you please tell me, is this possible to decrease the space of the virtual disk?

**Comment by vbevbe** [comment]
I've hit the same problem.

The new size updates correctly after I hit "consolidate" from the VM snapshots menu. Hope this help.
